pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +458,720 @@ pub mod pallet {
+							less_total_staked = Some(self.lowest_bottom_delegation_amount);
+						}
+						// insert into bottom
+						self.add_bottom_delegation::<T>(false, candidate, delegation);
+						DelegatorAdded::AddedToBottom
+					}
+				}
+				// top is either empty or partially full
+				_ => {
+					self.add_top_delegation::<T>(candidate, delegation);
+					DelegatorAdded::AddedToTop {
 						new_total: self.total_counted,
-					})
+					}
+				}
+			};
+			Ok((delegator_added, less_total_staked))
+		}
+		/// Add delegation to top delegation
+		/// Returns Option<negative_total_staked_remainder>
+		/// Only call if lowest top delegation is less than delegation.amount || !top_full
+		pub fn add_top_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegation: Bond<T::AccountId, BalanceOf<T>>,
+		) -> Option<Balance>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let mut less_total_staked = None;
+			let mut top_delegations = <TopDelegations<T>>::get(candidate)
+				.expect("CandidateInfo existence => TopDelegations existence");
+			let max_top_delegations_per_candidate = T::MaxTopDelegationsPerCandidate::get();
+			if top_delegations.delegations.len() as u32 == max_top_delegations_per_candidate {
+				// pop lowest top delegation
+				let new_bottom_delegation = top_delegations.delegations.pop().expect("");
+				top_delegations.total = top_delegations
+					.total
+					.saturating_sub(new_bottom_delegation.amount);
+				if matches!(self.bottom_capacity, CapacityStatus::Full) {
+					less_total_staked = Some(self.lowest_bottom_delegation_amount);
+				}
+				self.add_bottom_delegation::<T>(true, candidate, new_bottom_delegation);
+			}
+			// insert into top
+			top_delegations.insert_sorted_greatest_to_least(delegation);
+			// update candidate info
+			self.reset_top_data::<T>(&top_delegations);
+			self.delegation_count += 1u32;
+			<TopDelegations<T>>::insert(&candidate, top_delegations);
+			less_total_staked
+		}
+		/// Add delegation to bottom delegations
+		/// Check before call that if capacity is full, inserted delegation is higher than lowest
+		/// bottom delegation (and if so, need to adjust the total storage item accordingly fuck)
+		pub fn add_bottom_delegation<T: Config>(
+			&mut self,
+			bumped_from_top: bool,
+			candidate: &T::AccountId,
+			delegation: Bond<T::AccountId, BalanceOf<T>>,
+		) where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let mut bottom_delegations = <BottomDelegations<T>>::get(candidate)
+				.expect("CandidateInfo existence => BottomDelegations existence");
+			// if bottom is full, kick the lowest bottom (which is expected to be lower than input
+			// as per check)
+			let increase_delegation_count = if bottom_delegations.delegations.len() as u32
+				== T::MaxBottomDelegationsPerCandidate::get()
+			{
+				let lowest_bottom_to_be_kicked = bottom_delegations
+					.delegations
+					.pop()
+					.expect("if at full capacity (>0), then >0 bottom delegations exist; qed");
+				assert!(
```

Comment:
This assert is to handle the edge case in which the kicked lowest top delegation is equal to the lowest bottom delegation?

Reply:
Thanks, I removed it because it would panic in that case. IMO we still need to kick the lowest bottom in this case to enforce first come first served for equal delegation amounts.